### PR TITLE
Introduce NetworkMode option

### PR DESCRIPTION
**Pull request by pedro-r-marques**
Provides openshift origin with the ability to control the network settings of
sti containers.

### CODE REVIEW

**Comment by pedro-r-marques** [comment]
@rajatchopra This knob is used in order to propagate the network namespace from the builder Pod to the s2i docker container; this removes assumption on the networking settings on docker. The s2i docker pod can use the SDN solution used by the pod.

**Comment by pedro-r-marques** [comment]
The following diff shows the proposed usage of this knob:
https://github.com/openshift/origin/compare/master...pedro-r-marques:sti_network
I would like to submit that as a PR once the dependency from source-to-build is merged, assuming that the maintainers consider the proposal acceptable.

**Comment by rajatchopra** [comment]
The idea is helpful. But just asking the reverse question - is it too much hassle to be able to provide external network connectivity through docker's default network mode?

**Comment by pedro-r-marques** [comment]
@rajatchopra The access allowed to the builder can be different than the network access allowed to the node. e.g. builder should have access to the enterprise source control system. The node itself, probably not. I think this would be also useful to the redhat sdn plugin: from a practical standpoint, one would probably not want to have to manage and operate docker networking (docker bridge) plus SDN.
With this proposal the deployment template can set options to the builder that would control its network access: that is set on the netns of origin-pod and inherited by s2i. It seems to me that this is the desirable workflow: control all settings via deployment .json.

**Comment by rajatchopra** [comment]
Agreed, we had to go through hoops to accommodate docker's Linux bridge with our solution.
Do you mind putting this and the usage PR together in this one?

**Comment by pedro-r-marques** [comment]
@rajatchopra Pardon my cluelessness. How do i put the 2 PRs together ? The usage PR is in a separate git repository (origin). Which then copies this repository through its Godeps; i was under the impression that i would have to submit this PR; wait for it to get merged; update the Godeps on origin and then submit the PR on origin with Godeps + change.

File: pkg/api/types.go [hunk 1]
```diff
@@ -45,6 +45,10 @@ type Config struct {
 	// previous image from private repositories
 	IncrementalAuthentication docker.AuthConfiguration
 
+	// DockerNetworkMode is used to set the docker network setting to --net=container:<id>
+	// when the builder is invoked from a container.
+	DockerNetworkMode string
```
Comment (bparees) [inline pkg/api/types.go]: is there a set of valid values that can be defined in the api as constants?

**Comment by bparees** [comment]
@pedro-r-marques @rajatchopra will this also provide the builder-pods DNS settings to the s2i container?  specifically i'm wondering about this issue:
https://trello.com/c/f3azg2K1/685-builder-containers-don-t-have-openshift-cluster-dns
https://github.com/openshift/origin/issues/4303

**Comment by pedro-r-marques** [comment]
@bparees This is required in order to fix the DNS issue, when any network solution other than what is provided by docker is used. However the patch doesn't add the resolv.conf file from the parent Pod (which has the correct DNS settings). Would you like me to add this functionality through this patch or a separate PR ?

File: pkg/api/types.go (hunk 1 above)
Comment (pedro-r-marques) [inline pkg/api/types.go]: Added a validator to check allowed values.

**Comment by bparees** [comment]
Separate PR would be great. Thanks!

Ben Parees | OpenShift

**Comment by rajatchopra** [comment]
LGTM, @bparees please mark merge if you are okay.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -18,14 +19,34 @@ func ValidateConfig(config *api.Config) []ValidationError {
...
+	}
+	return false
+}
+
 // NewFieldRequired returns a *ValidationError indicating "value required"
 func NewFieldRequired(field string) ValidationError {
 	return ValidationError{ValidationErrorTypeRequired, field}
 }
 
+func NewFieldInvalidValue(field string) ValidationError {
```
Comment (mfojtik) [inline pkg/api/validation/validation.go]: need godoc

**Comment by mfojtik** [comment]
LGTM, please add the godoc and it is ready for merge

**Comment by pedro-r-marques** [comment]
@mfojtik Apologies for the newbie question... The PR has a comment in api/types.go type Config that is picked up when i run godoc ./pkg/api

output from godoc:
```
    // DockerNetworkMode is used to set the docker network setting to --net=container:<id>
    // when the builder is invoked from a container.
    DockerNetworkMode string
```

Can you please clarify your comment ? Are you referring to other documentation that is missing ?

**Comment by pedro-r-marques** [comment]
@mfojtik The godoc has been added.

File: pkg/api/types.go (hunk 1 above)
Comment (bparees) [inline pkg/api/types.go]: i still think you should define constants here for this.

**Comment by pedro-r-marques** [comment]
@bparees Added DockerNetworkMode and corresponding string constants.

**Comment by bparees** [comment]
thanks, lgtm.  [merge]

**Pull request merged**